### ISSUE TITLE
[Idea] `visible` assertion considers color contrasts

### ISSUE THREAD

**Issue by jorgelainfiesta** [thefrontside/bigtest#684]
## Motivation

The premise of Interactos is to enable developers to test their apps as a user would use them. We already have pretty good filters and there's many ways to write tests that check for accessibility.  In a later iteration of Interactors, we can take accessibility one step forward and bake color contrast in the visible scheme.

For instance, if an assertion says there should be a visible button with text "Submit", and it is there on the DOM but the color contrast is not enough then BIgTest would fail the assertion.

I think this feature would give a powerful argument to a unique selling point that we have: accessibility at the core. I can imagine landing pages, talks, and articles about Interactors once we make this narrative more robust.

It is also in line with our Leitmotif: work less, test more.

## Approach

I'm not sure how we can assess color contrast, but it must be done over the rendered element to account for background-images. It wouldn't be enough to compare `background` vs `color`. I've experienced false positives with tools that check color contrasts with that approach.

This could be an opt-in setup, and we can allow the user to select how restrictive they want their assertions to be. For example they could enable only `AAA` contrast, or be ok with `A` contrast. We can even issue a little warning when they set the contrast setting too low.

**Comment by cowboyd** [comment]
> work less, test more.

Man, why isn't this, or some some variation of this,  in our marketing material?

"do less, test more."
"slog less, test more."
"test more, for less."

**Comment by cowboyd** [comment]
To the idea, perhaps we could have a separate filter, like `constrastable: true` which would be default true if you have this feature enabled? We also may have to implement filter dependencies, like don't check the constrast if it is not visible.

**Comment by jorgelainfiesta** [comment]
> > work less, test more.
> 
> Man, why isn't this, or some some variation of this, in our marketing material?
> 
> "do less, test more."
> "slog less, test more."
> "test more, for less."

Because it's too vague for people to connect with it. It doesn't feel real or trustworthy. For us it's great because we know what we're trying to achieve, but for an outside it's just 'meh somebody else trying to sell me the sky'.

**Comment by jnicklas** [comment]
I think going by something like the parent elements background color is too simple, there are many CSS shenanigans which could affect this. So we'd have to do something like sampling color next to the element, but even that would be very error prone, since we'd have to sample a lot and then average it or something. Is sampling at a position even possible? I think this would be very difficult to implement.

It might be worth looking at some accessibility checkers and seeing how they implement similar functionality, though I suspect that those are mostly written as browser extensions.
